[WIP] Upgraded to Play 2.6 and Akka 2.5

jroper opened the pull request:
I've opened this PR so that the work I'm doing is visible.

This isn't yet finished, there are some major binary compatibility issues with Cassandra, and I've raised some fixes with akka-persistence-cassandra to address them, including the ability to use a shaded version of Cassandra with shaded dependencies.

ignasi35 commented on service/core/api/src/main/scala/TrampolineContextAccessor.scala:
```diff
@@ -0,0 +1,19 @@
+// [lagom].
+package play.lagom {
+  sealed trait Execution {
+    implicit def trampoline: ExecutionContext = play.core.Execution.trampoline
```
had to lookup what a trampoline is. :-(

jroper commented:
In the latest commit that I've added, I've been able to remove almost 50 dependencies from Lagom's whitelist.  Included in this list are things like spring and hibernate validator, which can be removed because play-java has separated out its form support from the core, Apache http client and signpost oauth, removed because play-ws' oauth support is now a separate module, asynchttpclient, removed because play-ws now uses a shaded asynchttpclient, and probably the biggest one is all of cassandras dependencies, which could be removed due to https://github.com/akka/akka-persistence-cassandra/pull/196.

TimMoore commented on dev/service-registry/client-javadsl/src/main/java/com/lightbend/lagom/internal/javadsl/registry/RegisteredService.java:
```diff
@@ -26,9 +22,7 @@
  */
 @SuppressWarnings("all")
-@ParametersAreNonnullByDefault
 @Generated({"Immutables.generator", "AbstractRegisteredService"})
```
Is this still being generated from Immutables? Were the annotation changes an unexpected side-effect?

jroper replied:
To be honest, I think we should replace these classes with hand written ones.

As I understand it, Immutables looks at your classpath, works out what annotations are present, and then uses them if they are present.  These annotations I think come from the findbugs javax annotations jar, which is no longer on the classpath after the Play 2.6 upgrade, and I didn't think it worth adding them just so these classes could still use them.  So I just deleted the annotations.

TimMoore replied:
> To be honest, I think we should replace these classes with hand written ones.

That sounds great.

I didn't realize that the `javax.annotation` package is split between the standard library and other jars! That seems like a really bad idea. I think Jigsaw might even disallow it.

jroper replied:
When I looked at findbugs 6 years ago, it wasn't under very active development.  I suspect that's still the case, so no ones really had any interest in solving the problem.

benmccann commented:
Will this PR be merged into master or a separate branch? I have a PR I'd like to send at some point, but it will depend on Play 2.6, so I was curious what the thinking was on it

dr3s commented:
Any thoughts on the serialization and other changes for applying a rolling update with lagom?  It sounds like we would need to perform two updates with the first one setting some config to 2.4 defaults before moving to the 2.5 defaults.

edouardKaiser commented on service/javadsl/integration-client/src/main/java/com/lightbend/lagom/javadsl/client/integration/LagomClientFactory.java:
```diff
@@ -187,12 +184,9 @@ public static LagomClientFactory create(String serviceName, ClassLoader classLoa
-        WSClient wsClient = new AhcWSClient(configBuilder.build(), materializer);
+        WSClientConfig wsClientConfig = new WSConfigParser(configuration.underlying(), environment.classLoader()).parse();
+        AhcWSClientConfig ahcWSClientConfig = new AhcWSClientConfigParser(wsClientConfig, configuration.underlying(), environment.classLoader()).parse();
+        WSClient wsClient = AhcWSClient.apply(ahcWSClientConfig, scala.Option.empty(), materializer);
```
The Play WS standalone lib offers the possibility to plug HTTP Caching, (`Option.emtpy()` here).
Could we provide a way to enable it?

edouardKaiser replied:
Also, I'm not sure about this one, but it looks like the `WsClient` injected in `JavadslServiceClientImplementor` hasn't changed in your diff. So it's still using the one from Play. (not sure why they kept a `play-ahc-ws` module in the framework instead of using the Play WS Standalone dependency, you might know better than me).

ignasi35 commented:
Trying to reproduce the failure in Travis I ran `kafka-broker-javadsl:test` and got a test failure on "should self-heal if failure occurs while running the publishing stream" which then causes the whole `test` task to hang. When that happens I have to reboot `sbt`.

jroper commented:
The error you saw is the same one that Travis sees (I just turned on logging), though the Travis test run isn't hanging when it fails.  It's really hard to know what the problem is - it is quite possible that the test is invalid - ie, we can't guarantee that message publishing will self heal within 30 seconds.

TimMoore commented on project/Dependencies.scala:
```diff
@@ -4,21 +4,21 @@ import sbt.Keys._
-  val AkkaStreamKafkaVersion = "0.13"
+  val NettyVersion = "4.1.8.Final"
+  val KafkaVersion = "0.10.2.0"
+  val AkkaStreamKafkaVersion = "0.15"
```
They have since released 0.16 https://github.com/akka/reactive-kafka/milestone/16?closed=1

Though it isn't clear to me whether any of the fixes affect Lagom.

ignasi35 commented:
> The error you saw is the same one that Travis sees (I just turned on logging), though the Travis test run isn't hanging when it fails. It's really hard to know what the problem is - it is quite possible that the test is invalid - ie, we can't guarantee that message publishing will self heal within 30 seconds.

The test flaky-ness can be reproduced locally both in sbt and intelliJ launching the `JavadslKafkaApiSpec` test when in this PRs branch. But the tests runs successfully when run in `master`.

I suspect the thread pool where this await [`messageCommitFailureInjected.await()`](https://github.com/lagom/lagom/blob/666e9b765725f18db099cc8db3dc1e6d22a58dbc/service/javadsl/kafka/server/src/test/scala/com/lightbend/lagom/internal/javadsl/broker/kafka/JavadslKafkaApiSpec.scala#L112-L112) happens may have different default settings causing the issue.

Another possible cause for error is the underlying Kafka/ZK not being fully cleaned-up.  I've noticed that often the first run after a certain delay will succeed but several runs in succession fail.

benmccann commented on project/Dependencies.scala:
```diff
@@ -4,21 +4,21 @@ import sbt.Keys._
-  val NettyVersion = "4.0.42.Final"
-  val KafkaVersion = "0.10.0.1"
-  val AkkaStreamKafkaVersion = "0.13"
+  val NettyVersion = "4.1.8.Final"
```
play recently upgraded to 4.1.10.Final

benmccann commented on project/Dependencies.scala:
```diff
@@ -4,21 +4,21 @@ import sbt.Keys._
-  val GuavaVersion = "19.0"
+  val AkkaPersistenceCassandraVersion = "0.52"
+  val ScalaTestVersion = "3.0.3"
+  val JacksonVersion = "2.8.7"
```
play is on 2.8.8

benmccann commented on project/Dependencies.scala:
```diff
@@ -4,21 +4,21 @@ import sbt.Keys._
   // Version numbers
-  val PlayVersion = "2.5.13"
-  val AkkaVersion = "2.4.17"
+  val PlayVersion = "2.6.0-M3"
```
2.6.0-R1 is out now. If you upgrade to it you will no longer need Lagom's build-link project and hopefully not its `Reloader` either

jroper commented:
The Kafka tests should now pass (assuming my assessment of why they were failing was true). Assuming they do, this should be merged.

Note that a number of the dependencies are now out of date (most notably, Play), but I think those should be updated in another pull request, this PR has been outstanding for too long.

jroper commented:
Well, travis failed, so obviously the Kafka tests aren't fixed.

ignasi35 commented:
After implementing the explicit creation of the topics I tried to run the test _without_ setting a consumer groupId per test: both changes are required.

I committed early (without changing the javadsl code to run the test on travis.

ignasi35 commented:
I'm manually running https://travis-ci.org/lagom/lagom/builds/236376455?utm_source=github_status&utm_medium=notification (may fail because it only has scaladsl fixed)
and https://travis-ci.org/lagom/lagom/builds/236382123?utm_source=github_status&utm_medium=notification (should pass because it has both scala and javadsl's fixed)
several times. So far results are consistent with expectations.

ignasi35 commented:
This branch has conflicts with master now (caused by d892c28e142ded7098993546f642f20bb940d95f )

benmccann commented on project/Dependencies.scala:
```diff
@@ -66,109 +71,75 @@ object Dependencies {
-      "com.datastax.cassandra" % "cassandra-driver-core" % "3.1.4",
-      "com.fasterxml" % "classmate" % "1.3.0",
+      "com.datastax.cassandra" % "cassandra-driver-core" % "3.2.0",
       "com.fasterxml.jackson.module" % "jackson-module-parameter-names" % JacksonVersion,
```
This module has become part of Jackson Java 8 Modules as of Jackson 2.8.5

TimMoore commented on persistence-cassandra/core/src/main/resources/lagom-test-embedded-cassandra.yaml:
```diff
@@ -25,6 +25,8 @@ saved_caches_directory: $DIR/saved_caches
 hints_directory: $DIR/hints
 
+cdc_raw_directory: $DIR/cdc_raw
+
```
Think we can remove this file and revert to the default config shipped with Akka Persistence Cassandra.

rcavalcanti commented:
I updated this so we can merge it onto lagom:play-2.6-akka-2.5 branch.

Since it's the first time we are doing such a thing (at least me), I did a backup of it locally.

rcavalcanti commented:
Merging that using GitHub is quite of a challenge. I will simply merge locally and push James' original branch directly into lagon:play-2.6-akka-2.5

rcavalcanti commented:
This PR was merged onto lagom:play-2.6-akka-2.5 and new PR is opened (#808)

ignasi35 commented:
See #466 and http://doc.akka.io/docs/akka/current/scala/project/migration-guide-2.4.x-2.5.x.html#additional-serialization-bindings
